---
title: "Process raw data"
author: "David Wu"
output: html_notebook
---
## Purpose
Process raw data. Provided data is already processed by cellranger; use this notebook to organize metadata

## Setup
### Use the here package for robust relative paths
```{r}
library(here)
```

### Optionally convert this Rmd to R script 
```{r eval=FALSE, message=FALSE}
current_file <- rstudioapi::getActiveDocumentContext()$path
output_file <- stringr::str_replace(current_file, '.Rmd', '.R')
knitr::purl(current_file, output = output_file)
file.edit(output_file)
```

### Directories
```{r, message=FALSE}
analysis_dir <- here('analysis/output/00_process') # analysis file output directory
data_dir <- here('data/derived/pipestance') # data file output directory

dir.create(analysis_dir, showWarnings = FALSE)
dir.create(data_dir, showWarnings = FALSE)
```

### Load libraries
```{r, message=FALSE}
library(tidyverse)
```

### Copy pipestance files from Box
Start in project directory
```{bash eval=FALSE} 
cd data/derived
rclone sync -P box:/Files pipestance
```

## Analysis
### Metadata table
```{r}
metadata <- read_tsv(here('analysis/input/metadata.tsv'))
```

### Examine metadata
```{r}
metadata
```
```{r}
metadata %>% filter(`Skin id #` !=  `Skin #`) %>% select(contains('id'))
```

### Prepare metadata subset
```{r}
dest_dir <- 'analysis/input/samples/'
dir.create(dest_dir)

metadata_subset <- 
  metadata %>% 
  mutate(id = paste0('S', id)) %>% 
  filter(is.na(exclude),
         tildra == 'Yes') %>% 
  select(sample,
         patient,
         id,
         condition,
         treatment,
         response,
         group,
         sample_label,
         patient_label,
         fastq,
         pipestance,
         adt,
         assay,
         published,
         citeseq = `Awaiting Cite Seq?`,
         flow = `CITE-seq (pre or post-flow antibody addition)`,
         chemistry = `10x kit`,
         n_cells = `est # of cells`,
         n_reads = `RNA total reads`,
         reads_per_cell = `mean reads/cell`,
         genes_per_cell = `median genes per cell`,
         saturation = `sequencing saturation`,
         f_ab_reads = `fraction αbody reads`,
         f_ab_reads_usable = `fraction αbody reads usable`,
         f_ab_unrecog = `fraction unrecognized αbody`,
         f_ab_reads_in_cells = `αbody reads in cells`,
         n_ab_reads_usable_per_cell = `αbody reads usable/cell`,
         mouse = `mouse spike in?`,
         reorder) %>% 
  arrange(reorder) %>% 
  mutate(adt = ifelse(is.na(f_ab_reads), 'No', 'Yes'),
         mouse = ifelse(mouse == 'yes', 'Yes', 'No'),
         pipestance = paste0('data/derived/pipestance/', pipestance),
         h5 = paste0(dest_dir, id, '.h5'),
         sample_order = row_number()) 

metadata_subset
```

### Match to output directories
```{r}
paths <- list.files(data_dir)
paths
```
### Check for missing paths
```{r}
metadata_subset[!(basename(metadata_subset$pipestance) %in% paths),]
```

```{r}
paths[!(paths %in% metadata$`Box folder Name`)]
```
```{r}
metadata_subset
```

### Transfer .h5 files to input directory
```{r}
transfer <- metadata_subset %>% 
  filter(basename(pipestance) %in% paths) %>% 
  select(sample, id, pipestance, h5)

transfer
```
```{r}
file.copy(from = here(transfer$pipestance, '/filtered_feature_bc_matrix.h5'),
          to = here(transfer$h5), 
          overwrite = TRUE)
```


## Export
```{r}
metadata_subset %>% filter(basename(pipestance) %in% paths) %>% write_tsv(here(analysis_dir, 'metadata_subset.tsv'))
```

## Session info
```{r}
sessionInfo()
```

