---
title: "Process Visium Data"
author: "Abed Kurdi and David Wu"
---


# integrating single-cell spatial RNA-seq data

Ref1: https://nbisweden.github.io/workshop-scRNAseq/labs/compiled/seurat/seurat_07_spatial.html#Identification_of_Spatially_Variable_Features
Ref2: https://satijalab.org/seurat/articles/spatial_vignette.html#working-with-multiple-slices-in-seurat-1

## Setup 
Set working directory to project directory

```{r setup}
require(knitr)
opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

### Optionally convert this Rmd to R script 
```{r eval=FALSE, message=FALSE}
current_file <- rstudioapi::getActiveDocumentContext()$path
output_file <- stringr::str_replace(current_file, '.Rmd', '.R')
knitr::purl(current_file, output = output_file)
file.edit(output_file)
```

### Directories
```{r, message=FALSE}
output_dir <- 'analysis/output/10_process_visium' # analysis file output directory
data_dir <- 'data/derived/seurat' # data file output directory
samples_dir <- "analysis/input/visium_samples/"

dir.create(output_dir, showWarnings = FALSE)
dir.create(data_dir, showWarnings = FALSE)
```

### Loading the libraries
```{r, message=FALSE}
library(Seurat) 
library(tidyverse)
library(ggExtra)
library(SeuratWrappers)
library(wutilities) # devtools::install_github('symbiologist/wutilities')
library(tictoc)
library(ggforce)
library(patchwork)
```

### Metadata
```{r}
metadata <- read_tsv("analysis/output/00_process/metadata_subset.tsv")
visium_sample_data <- read_tsv('analysis/input/visium_sample_data.tsv') %>% drop_na() %>% mutate(id = ifelse(id == 'skin234', 'skin277', id))
```
#### Unify metadata
```{r}
metadata_df <- metadata %>% 
  inner_join(visium_sample_data)

metadata_df
```


```{r}
all_samples <- list()

# I renamed "tissue_positions.csv" to "tissue_positions_list.csv"
# load in the samples and remove genes with all zero counts: https://github.com/satijalab/sctransform/issues/86
for(s in dir(samples_dir)){
	print(s)
	all_samples[[s]] <- Load10X_Spatial(
	  data.dir = paste0(samples_dir,s), 
	  filename = "filtered_feature_bc_matrix.h5")
	
	all_samples[[s]] <- subset(all_samples[[s]], subset = nCount_Spatial > 0)
	
	all_samples[[s]]@images$slice1@coordinates <- all_samples[[s]]@images$slice1@coordinates %>% mutate_all(as.integer)
	
}
```
# add metadata
```{r}
metadata_to_add <- 
  c('sample',
    'id',
    'CoreID',
    'treatment',
    'response')

for(n in names(all_samples)){
	md <- subset(metadata_df, metadata_df$id == n)
	
	for(m in metadata_to_add){
		print(paste0("adding ",m))
		all_samples[[n]][[m]] <- md[[m]]
	}
}

# reorder
reordered <- metadata_df %>% arrange(CoreID) %>% pull(id)
all_samples <- all_samples[reordered]
```
### Log transform without normalization
```{r}
#all_samples <- all_samples %>% map(NormalizeData)

all_samples <- all_samples %>% map(function(i) {
  
  i@assays$Spatial@data <- as.matrix(log2(i@assays$Spatial@counts + 1))
  
  i
  
})
```

```{r}
SpatialFeaturePlot(all_samples$skin277, features = 'nFeature_Spatial')
```


```{r}
SpatialFeaturePlot(
  all_samples$skin277, features = c("KRT10", 'COL1A1'), image.alpha = 0) 
```
```{r}
p <- all_samples %>% map(function(i) {
  
  complexity <- i$nFeature_Spatial %>% median()
  SpatialFeaturePlot(i, features = c("KRT10"), image.alpha = 0) + 
    labs(title = paste0(i$CoreID %>% unique() %>% str_remove('6355-'),': ', i$id %>% unique()), 
         subtitle = paste0('Median genes: ', complexity)) +
    theme(plot.background = element_rect(color = 'black'))
}) %>% wrap_plots(ncol = 4)

p
  
```

```{r}
p <- all_samples %>% map(function(i) {
  
  complexity <- i$nFeature_Spatial %>% median()
  SpatialFeaturePlot(i, features = c("COL1A1"), image.alpha = 0) + 
    labs(title = paste0(i$CoreID %>% unique() %>% str_remove('6355-'),': ', i$id %>% unique()), 
         subtitle = paste0('Median genes: ', complexity)) +
    theme(plot.background = element_rect(color = 'black'))
}) %>% wrap_plots(ncol = 4)

p
```


```{r}
p <- all_samples %>% map(function(i) {
  
  complexity <- i$nFeature_Spatial %>% median()
  SpatialFeaturePlot(i, features = c("nFeature_Spatial"), image.alpha = 0) + 
    labs(title = paste0(i$CoreID %>% unique() %>% str_remove('6355-'),': ', i$id %>% unique()), 
         subtitle = paste0('Median genes: ', complexity)) +
    theme(plot.background = element_rect(color = 'black'))
}) %>% wrap_plots(ncol = 4)

p
```

### Normalized analysis
```{r}
all_samples_norm <- all_samples %>% map(NormalizeData)
```

```{r}
p <- all_samples_norm %>% map(function(i) {
  
  complexity <- i$nFeature_Spatial %>% median()
  SpatialFeaturePlot(i, features = c("KRT10"), image.alpha = 0) + 
    labs(title = paste0(i$CoreID %>% unique() %>% str_remove('6355-'),': ', i$id %>% unique()), 
         subtitle = paste0('Median genes: ', complexity)) +
    theme(plot.background = element_rect(color = 'black'))
}) %>% wrap_plots(ncol = 4)

p
  
```

```{r}
p <- all_samples_norm %>% map(function(i) {
  
  complexity <- i$nFeature_Spatial %>% median()
  SpatialFeaturePlot(i, features = c("COL1A1"), image.alpha = 0) + 
    labs(title = paste0(i$CoreID %>% unique() %>% str_remove('6355-'),': ', i$id %>% unique()), 
         subtitle = paste0('Median genes: ', complexity)) +
    theme(plot.background = element_rect(color = 'black'))
}) %>% wrap_plots(ncol = 4)

p
```


```{r}
p <- all_samples_norm %>% map(function(i) {
  
  complexity <- i$nFeature_Spatial %>% median()
  SpatialFeaturePlot(i, features = c("nCount_Spatial"), image.alpha = 0) + 
    labs(title = paste0(i$CoreID %>% unique() %>% str_remove('6355-'),': ', i$id %>% unique()), 
         subtitle = paste0('Median genes: ', complexity)) +
    theme(plot.background = element_rect(color = 'black'))
}) %>% wrap_plots(ncol = 4)

p
```



### SCT analylsis

```{r}

all_samples_sct <- all_samples %>% map(SCTransform, assay = 'Spatial')
```




```{r}
# run SCT on all datasets
all_samples <- all_samples %>% map(function(i) {
  
  print2('Running sample: ', i$id %>% unique())
  SCTransform(i, assay = "Spatial")
  
}) 
```

```{r}
p <- all_samples %>% map(function(i) {
  
  complexity <- i$nFeature_Spatial %>% median()
  spots <- ncol(i)
  
  SpatialFeaturePlot(i, features = c("KRT10"), image.alpha = 0.1) + 
    labs(title = i$id %>% unique(), 
         subtitle = paste0('Median genes: ', complexity, ', Spots: ', spots)) +
    theme(plot.background = element_rect(color = 'black'))
}) %>% wrap_plots(ncol = 4)

p
```
## skin292
```{r}
p_krt <- SpatialFeaturePlot(all_samples$skin292, features = c("KRT10", 'COL1A1'), image.alpha = 0.1) + 
    labs(title = 'skin292') 

p_krt
```
```{r}

```


#### Integration
```{r}
# need to set maxSize for PrepSCTIntegration to work
options(future.globals.maxSize = 2000 * 1024^2)  # set allowed size to 2K MiB
```


```{r}
# select integration features and run SCTIntegration
features = SelectIntegrationFeatures(all_samples, nfeatures = 3000, verbose = FALSE)
all_samples <- PrepSCTIntegration(object.list = all_samples, anchor.features = features,
    verbose = FALSE)

# samples integration
int.anchors <- FindIntegrationAnchors(object.list = all_samples, normalization.method = "SCT",
    verbose = FALSE, anchor.features = features)
samples_integrated <- IntegrateData(anchorset = int.anchors, normalization.method = "SCT",
    verbose = FALSE)
```


```{r}
# dimensionality reduction and clustering
samples_integrated <- RunPCA(samples_integrated, verbose = FALSE)
samples_integrated <- FindNeighbors(samples_integrated, dims = 1:30)
samples_integrated <- FindClusters(samples_integrated, resolution = seq(0.1,2,by=0.1), verbose = FALSE)
samples_integrated <- RunUMAP(samples_integrated, dims = 1:30)
```


```{r}
p <- SpatialFeaturePlot(samples_integrated, features = 'KRT10', ncol = 4)
```


```{r}
# (!) to choose one image (or slice): https://github.com/satijalab/seurat/issues/5774
# p1 <- SpatialDimPlot(samples_integrated, images = c("Skin234","Skin244"))+NoLegend()
# p2 <- SpatialDimPlot(samples_integrated, images = c("skin296","skin269"))+NoLegend()
# p3 <- SpatialDimPlot(samples_integrated, images = c("Skin235","skin279"))+NoLegend()
# p4 <- SpatialDimPlot(samples_integrated, images = c("skin285","skin292"))+NoLegend()
# p5 <- SpatialDimPlot(samples_integrated, images = c("skin289","skin290"))+NoLegend()
# p6 <- SpatialDimPlot(samples_integrated, images = c("skin305","skin304"))+NoLegend()

# p1 / p2 / p3 / p4 / p5 / p6

Idents(samples_integrated) <- "integrated_snn_res.1"

pdf(paste0("analysis/output/10_visium/UMAP_integrated_per_cluster_and_treatment.pdf"), height=8, width=14)
DimPlot(samples_integrated, reduction = "umap", group.by = c("ident", "treatment"))
dev.off()

for(n in unique(samples_integrated$id)){
	pdf(paste0("analysis/output/10_visium/",n,"_spatial_umap.pdf"), height=10, width=10)
	print(SpatialDimPlot(samples_integrated, images = n, alpha=1.5))
	dev.off()
}

# DE between pre vs post - resolution 1
clusters <- as.character(unique(Idents(samples_integrated)))

OUT <- createWorkbook()
for(cluster in clusters){
	tryCatch({
	print(paste0(cluster))
	markers <- FindMarkers(samples_integrated, ident.1=cluster, logfc.threshold=0, 
	min.pct=0.01, test.use="MAST") %>% 
	mutate(., gene = rownames(.))

	addWorksheet(OUT, paste0(cluster))
	writeData(OUT, sheet = paste0(cluster), x = markers)
	}, error=function(e){message("Cell group 1 has fewer than 3 cells")})
}

saveWorkbook(OUT, paste0("analysis/output/11_visium_de/marker_genes_per_cluster.xlsx"))

# Annotation based on the markers above - marker_genes_per_cluster.xlsx
new.cluster.ids <- c("Fibroblast","Keratinocytes","Fibroblast","Endothelial","Fibroblast","Fibroblast",
	"Fibroblast","Fibroblast","Fibroblast","Keratinocytes","Fibroblast","Fibroblast",
	"Keratinocytes","Keratinocytes","Fibroblast","Fibroblast","Keratinocytes","Keratinocytes")

samples_integrated$annotation <- Idents(samples_integrated)

names(new.cluster.ids) <- levels(samples_integrated)
samples_integrated <- RenameIdents(samples_integrated, new.cluster.ids)

samples_integrated$anno_res_1 <- Idents(samples_integrated)

# UMAP with annotation
pdf("analysis/output/10_visium/UMAP_with_annotation.pdf",width=8, height=8)
DimPlot(samples_integrated)
dev.off()

# save the object
#saveRDS(samples_integrated, "/media/pk3/143F2A3651271F75/special_projects/8_single_cell_spatial_jeffrey/integrated_samples.rds")


# extract data - as per David's request
## get the donor, treatment and cluster columns for all cells
table1 <- samples_integrated@meta.data[,c("id","treatment","integrated_snn_res.1","anno_res_1")]
table1 <- data.frame(cell_barcode = rownames(table1), donor=table1$id, treatment=table1$treatment, cluster=table1$integrated_snn_res.1, annotation=table1$anno_res_1)
write.table(table1, "analysis/output/10_visium/cell_barcodes_donor_treatment_and_cluster_at_res_1.txt", 
	col.names=TRUE, row.names=FALSE, quote=FALSE, sep='\t')

## get the normalized counts for the genes of interest
normalized_counts_specific_genes <- as.matrix(samples_integrated@assays$SCT@data) %>% subset(., rownames(.) %in% c("IL17A","IL17F","IL23A","IL23R","ZFP36L2","ZFP36"))
normalized_counts_specific_genes <- cbind(genes=rownames(normalized_counts_specific_genes), normalized_counts_specific_genes)
rownames(normalized_counts_specific_genes) <- NULL
normalized_counts_specific_genes <- as.data.frame(normalized_counts_specific_genes)

write.table(normalized_counts_specific_genes, "analysis/output/10_visium/normalized_counts_for_genes_of_interest.txt", 
	col.names=TRUE, row.names=FALSE, quote=FALSE, sep='\t')

```

