---
title: "Secukinumab"
author: "David Wu"
---

Set working directory to project directory
```{r setup}
library(here)
```

### Optionally convert this Rmd to R script 
```{r eval=FALSE, message=FALSE}
current_file <- rstudioapi::getActiveDocumentContext()$path
output_file <- stringr::str_replace(current_file, '.Rmd', '.R')
knitr::purl(current_file, output = output_file)
file.edit(output_file)
```

### Directories
```{r, message=FALSE}
output_dir <- 'analysis/output/secukinumab' # analysis file output directory
data_dir <- 'data/derived/tildra' # data file output directory

dir.create(output_dir, showWarnings = FALSE)
dir.create(data_dir, showWarnings = FALSE)
```

### Loading the libraries
```{r, message=FALSE}
library(Seurat) 
library(tidyverse)
library(ggExtra)
library(ggthemes)
library(ggforce)
#library(SeuratWrappers) # devtools::install_github('satijalab/seurat-wrappers')
library(wutilities) # devtools::install_github('symbiologist/wutilities')
library(tictoc)
library(rcartocolor)
library(ggrepel)
library(rtracklayer)
library(plyranges)
theme_set(theme_dwu()) # set default theme

```

### Data from Liu et al. https://www.jidinnovations.org/article/S2667-0267(21)00096-5/fulltext#secsectitle0065
GEO link: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE171012
```{r}
secu <- read_tsv(here('analysis/input/GSE171012_CountData_SecukinumabPsoRnaSeq_20210324.tsv.gz'))

secu
```

```{r}
load(here('data/external/20180706_Novartis062518_FilteringAndQC/RObj-CountAndAnnoTables_Novartis062518_20190624.rdata'))
```

```{r}
anno.n062518
```

```{r}
anno.n062518.final
```

### Gene signature
```{r}
ray_genes <- c("IL17F", "GNLY", "LRRN3", "MGAT4A", "CHN1", "CTSH", "PTPN13", "GZMB", "ADGRG1", "SOX4", "GABPB1-AS1", "SNX9", "CD7", "CPM", "CTLA4", "ODF2L", "CBLB", "LAYN", "TNFAIP3", "ARHGEF12", "H1FX", "HIST1H1E", "YPEL2", "GBP5", "DAPK2", "RAP1B", "CXCL13", "FYN")
```

## Data processing
```{r}
annotation <- import('data/external/liu_secukinumab/gencode.v25.annotation.gtf.gz')
ens2symbol <- annotation@elementMetadata %>% as_tibble() %>% filter(type == 'gene') %>% select(gene_id, gene_name) %>% unique()
```

```{r}
counts <- counttable.n062518
metadata <- anno.n062518 %>% rownames_to_column('sample') %>% 
  mutate(Response = case_when(
    str_detect(PatientOutcome.Week52,'PASI75|PASI100') ~ 'Responder',
    str_detect(PatientOutcome.Week52,'PASI50|limited') ~ 'Non-responder',
  ))

```
```{r}
metadata %>% 
  select(SubjectId, WeekGroup, Week, CellType, Status) %>% 
  summarize_all(n_distinct)
```

### Subjects with PSO at Week 0 and 12
```{r}
metadata_subset <- metadata %>% 
  filter(QC_Passed,
         Week %in% c(0, 12),
         Status == 'PSO') %>% 
  select(sample, SubjectId, Week, CellType, PASI, PASI.Reduction.Week0_Week52, PatientOutcome.Week52, Response) %>% 
  group_by(SubjectId) %>% 
  add_count() %>% 
  arrange(SubjectId, Week, CellType)

metadata_subset
```

### Normalize
```{r}
tpm <- t(t(counts) / colSums(counts)) * 1e6
```


```{r}
logtpm <- log10(tpm + 0.01)
```

```{r}
gene_id_subset <- ens2symbol %>% filter(gene_name %in% ray_genes)
```

```{r}
gene_subset <- tpm[gene_id_subset$gene_id, ] %>% as.data.frame() %>% rownames_to_column('gene_id') %>% inner_join(gene_id_subset) %>% select(gene_id, gene_name, everything())
gene_subset
```

```{r}
expression_table_long <- gene_subset %>% 
  pivot_longer(cols = -starts_with('gene'),
               names_to = 'sample',
               values_to = 'expression') %>% 
  inner_join(metadata_subset) %>% 
  mutate(Treatment = factor(ifelse(Week == 0, 'Pre', 'Mid'), levels = c('Pre', 'Mid'))) %>% 
  select(-sample, -n) %>% 
  arrange(SubjectId)

expression_table_long
```
```{r}
expression_table_wide <- 
  expression_table_long %>% 
  select(SubjectId, CellType, gene_name, Treatment, expression, PatientOutcome.Week52, Response) %>% 
  pivot_wider(names_from = Treatment,
              values_from = expression) %>% 
  drop_na() %>% 
  arrange(SubjectId)
  
expression_table_wide
```

```{r}
plot_dir <- file.path(output_dir, 'waterfall')
dir.create(output_dir, showWarnings = FALSE)

waterfall_plots <- map(expression_table_long$SubjectId %>% unique(), function(i) {
  
  table_subset <- expression_table_long %>% 
    filter(SubjectId == i)
  response <- table_subset$Response %>% unique()
  pasi_week0 <- table_subset %>% filter(Treatment == 'Pre') %>% pull(PASI) %>% unique()
  pasi_week12 <- table_subset %>% filter(Treatment == 'Mid') %>% pull(PASI) %>% unique()
    
  p <- table_subset %>%  
   ggplot(aes(x = reorder(gene_name, -expression),
             y = expression,
             color = Treatment,
             group = gene_name)) +
  geom_point(size = 3) +
  geom_line(color = 'grey50', arrow = arrow(length = unit(0.2, "cm"), type = 'closed')) +
  theme_dwu(legend.position = 'top') +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5)) +
  scale_color_manual(values = c('orangered', 'dodgerblue4')) +
  #geom_point(size = 2, data = summarized_table %>% filter(treatment == 'Healthy'), inherit.aes = FALSE, aes(x = reorder(gene, -mean), y = mean)) +
  facet_wrap(~CellType, scales = 'free_y', nrow = 2) +
  labs(
    title = paste0('Patient ', i, ': ', response),
    subtitle = paste0('PASI week 0 = ', pasi_week0, ', week 12 = ', pasi_week12),
    x = '',
    y = 'log10 TPM expression')
  
  ggsave(plot = p,
         filename = paste0(response, '_', i, '.png'),
         path = plot_dir,
         h = 8,
         w = 10,
         dpi = 150)
  
  p

})
  
```

### Grouped analysis
```{r}
celltypes <- expression_table_long$CellType %>% unique() %>% sort()


boxplots <- map(celltypes, function(i) {

  expression_table_long %>% 
    filter(CellType == i) %>% 
    mutate(Response = factor(Response, levels = c('Responder', 'Non-responder'))) %>% 
  ggplot(aes(x = Response,
             y = expression,
             fill = Treatment)) +
  geom_boxplot(alpha = 0.5) +
  facet_wrap(~gene_name, nrow = 2, scales = 'free_y') +
  scale_fill_manual(values = c('orangered', 'dodgerblue4')) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5)) +
  theme(legend.position = 'top') +
  labs(title = i,
       x = '',
       y = 'Transcripts per million') 
})



  
```



## Pseudobulk
```{r}
seuratobj <- read_rds(file.path(data_dir, 'seuratobj_annotated.rds'))
meta <- seuratobj@meta.data %>% rownames_to_column() 
```


```{r}

```



### Extract expression
```{r}
genes_to_extract <- ray_genes %>% unique()

metadata_subset <- meta %>% 
  mutate(treatment = ifelse(treatment == 'None', 'Healthy', treatment),
         treatment = factor(treatment, levels = c('Pre', 'Mid', 'Healthy')))

patient_matrix <- map(metadata_subset$id %>% unique() %>% set_names(.), function(i) {
  
  patient_cells <- metadata_subset %>% filter(id == i) %>% pull(rowname)
  patient_counts <- rowSums(seuratobj@assays$RNA@counts[,patient_cells])
  
  
  patient_counts
}) %>% do.call(cbind, .)
```


```{r}
tpm <- t(t(patient_matrix) / colSums(patient_matrix)) * 1e6
```

###
```{r}
gene_subset <- tpm[genes_to_extract, ] %>% as.data.frame() %>% rownames_to_column('gene_name')
gene_subset
```

```{r}
expression_table_long <- gene_subset %>% 
  pivot_longer(cols = -starts_with('gene'),
               names_to = 'id',
               values_to = 'expression') %>% 
  inner_join(metadata_subset %>% select(id, patient, treatment, response_group) %>% unique()) 
  
  
expression_table_long
```
```{r}
expression_table_wide <- 
  expression_table_long %>% 
  select(SubjectId, CellType, gene_name, Treatment, expression, PatientOutcome.Week52, Response) %>% 
  pivot_wider(names_from = Treatment,
              values_from = expression) %>% 
  drop_na() 
  
expression_table_wide
```

```{r}
plot_dir <- file.path(output_dir, 'waterfall')
dir.create(output_dir, showWarnings = FALSE)

waterfall_plots <- map(expression_table_long$SubjectId %>% unique(), function(i) {
  
  table_subset <- expression_table_long %>% 
    filter(SubjectId == i)
  response <- table_subset$Response %>% unique()
  pasi_week0 <- table_subset %>% filter(Treatment == 'Pre') %>% pull(PASI) %>% unique()
  pasi_week12 <- table_subset %>% filter(Treatment == 'Mid') %>% pull(PASI) %>% unique()
    
  p <- table_subset %>%  
   ggplot(aes(x = reorder(gene_name, -expression),
             y = expression,
             color = Treatment,
             group = gene_name)) +
  geom_point(size = 3) +
  geom_line(color = 'grey50', arrow = arrow(length = unit(0.2, "cm"), type = 'closed')) +
  theme_dwu(legend.position = 'top') +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5)) +
  scale_color_manual(values = c('orangered', 'dodgerblue4')) +
  #geom_point(size = 2, data = summarized_table %>% filter(treatment == 'Healthy'), inherit.aes = FALSE, aes(x = reorder(gene, -mean), y = mean)) +
  facet_wrap(~CellType, scales = 'free_y', nrow = 2) +
  labs(
    title = paste0('Patient ', i, ': ', response),
    subtitle = paste0('PASI week 0 = ', pasi_week0, ', week 12 = ', pasi_week12),
    x = '',
    y = 'log10 TPM expression')
  
  ggsave(plot = p,
         filename = paste0(response, '_', i, '.png'),
         path = plot_dir,
         h = 8,
         w = 10,
         dpi = 150)
  
  p

})
  
```

### Grouped analysis
```{r}





  p <- expression_table_long %>% 
    filter(treatment != 'Healthy') %>% 
    mutate(response_group = factor(response_group, levels = c('Responder', 'Non-responder'))) %>% 
  ggplot(aes(x = response_group,
             y = expression,
             fill = treatment)) +
  geom_boxplot(alpha = 0.5) +
  facet_wrap(~gene_name, nrow = 2, scales = 'free_y') +
  scale_fill_manual(values = c('orangered', 'dodgerblue4', 'grey50')) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        plot.subtitle = element_text(hjust = 0.5)) +
  theme(legend.position = 'top') +
    labs(x = '',
         y = 'Transcripts per million',
         title = 'tildra')
  
  p




  
```

## Session Info
```{r}
sessionInfo()
```



